spirv_glsl.cpp

```diff
@@ -719,94 +719,51 @@ void CompilerGLSL::emit_header()
-			statement("#extension GL_ARB_shader_ballot : require");
-			statement("#extension GL_ARB_gpu_shader_int64 : require");
-			statement("#define gl_SubgroupEqMask uvec4(unpackUint2x32(gl_SubGroupEqMaskARB), 0u, 0u)");
-			statement("#define gl_SubgroupGeMask uvec4(unpackUint2x32(gl_SubGroupGeMaskARB), 0u, 0u)");
-			statement("#define gl_SubgroupGtMask uvec4(unpackUint2x32(gl_SubGroupGtMaskARB), 0u, 0u)");
-			statement("#define gl_SubgroupLeMask uvec4(unpackUint2x32(gl_SubGroupLeMaskARB), 0u, 0u)");
-			statement("#define gl_SubgroupLtMask uvec4(unpackUint2x32(gl_SubGroupLtMaskARB), 0u, 0u)");
+		else
+			statement("#extension ", ext, " : require");
+	}
 
-			statement("#else");
-			statement("#error No extensions available to emulate GL_KHR_shader_subgroup_ballot.");
+	if (!options.vulkan_semantics)
+	{
+		using Supp = KHR_shader_subgroupSupportHelper;
 
-			statement("#endif");
-		}
-		else if (ext == "GL_KHR_shader_subgroup_vote")
+		Supp::Result result = shader_subgroup_supporter.resolve();
+		for (uint32_t ft = 0u; ft < Supp::FeatureCount; ++ft)
 		{
-			declared_KHR_shader_subgroup_vote = true;
-			if (options.vulkan_semantics)
-				statement("#extension GL_KHR_shader_subgroup_vote : require");
-			else
+			Supp::Feature feature = static_cast<Supp::Feature>(ft);
+			if (!shader_subgroup_supporter.isFeatureRequested(feature))
+				continue;
+
+			auto exts = shader_subgroup_supporter.getCandidatesForFeature(feature, result);
+			if (exts.empty())
+				continue;
+
+			statement("#ifdef GL_KHR_shader_subgroup");
+			statement("#extension GL_KHR_shader_subgroup : require");
+			for (auto it = exts.begin(); it != exts.end(); ++it)
 			{
-				statement("#ifdef GL_KHR_shader_subgroup");
-				statement("#extension GL_KHR_shader_subgroup : require");
-				statement("#elif defined(GL_ARB_shader_group_vote)");
-				statement("#extension GL_ARB_shader_group_vote : require");
-				statement("#elif defined(GL_NV_gpu_shader5)");
-				statement("#extension GL_NV_gpu_shader5 : require");
-				statement("#elif defined(GL_AMD_gcn_shader) && defined(GL_AMD_gpu_shader_int64)");
-				statement("#extension GL_AMD_gcn_shader : require");
-				statement("#extension GL_AMD_gpu_shader_int64 : enable");
-				statement("#extension GL_NV_gpu_shader5 : enable");
-				statement("#else");
-				statement("#error No extensions available for subgroupAllEqual()");
-				statement("#endif");
+				std::string name = Supp::getExtensionName(*it);
+				auto extraNames = Supp::getExtraRequiredExtensionNames(*it);
+				if (extraNames.empty())
+					statement("#elif defined(", name, ")");
+				else
+				{
+					std::string extra;
+					for (const auto &e : extraNames)
+						extra += " && defined(" + e + ")";
+					statement("#elif defined(", name, ")", extra);
+				}
+				for (const auto &e : extraNames)
+					statement("#extension ", e, " : required");
+				statement("#extension ", name, " : required");
```

Comment:
how does this work, what is it supposed to emit?

Reply:
```
#elif defined(ext1) && defined(extra1)
#extension ext1 : required
#extension extra1 : required
#elif defined(ext2) && defined(extra2)
#extension ext2 : required
#extension extra2 : required
...
...
...
#else
#error
#endif
```

Reply:
not sure the `extra` extension thing is a good design idea. I'd rather have the candidate extension have its own array of statements for dependancy/ladder extensions.

I.e for `AMD_gcn_shader` this could be
```cpp
{
"(defined(GL_AMD_gpu_int64_t) || defined(GL_NV_gpu_shader5))",
"#extension GL_AMD_gpu_int64_t : enable",
"#extension GL_NV_gpu_shader5 : enable"
}
```

Because there's a lot of "either/or" relations and possibly extensions depending on extensions.

Reply:
yeah, i noticed before and already fixed